### ISSUE TITLE
Resize operations (override and sizeMultiplier) sometimes causes inexact fit into ImageView

### ISSUE THREAD

**Issue by xGary** [bumptech/glide#1621]
I am trying to use override() (also tried sizeMultiplier()) to downscale my images. However, in some cases, that causes the image to not fit exactly into the imageview leaving a very small gap. The gap will happen on the side that has dimension of MATCH_PARENT. My ImageView is using fitCenter, if I use fitXY, then the gap doesn't happen but that is not a solution since they're not the same thing.

The Glide load code is:

```java
Glide.with(context)
                     .load(imageResource)
                     .override(overrideWidth, Target.SIZE_ORIGINAL)
                     .into(imageView1);
```

And ImageView:

```xml
<ImageView
    android:id="@+id/imageview1"
    android:layout_width="match_parent"
    android:layout_height="wrap_content"
    android:layout_gravity="center"
    android:adjustViewBounds="true"
    android:scaleType="fitCenter"
    />
```

Not all override values cause the gap, so you may have to try a few values before it happens. I attached an APK that demonstrates this issue. The FAB loads a black image with Glide with a random override width value between 1-1000 and a override height of Target.SIZE_ORIGINAL. There is a runnable that rapidly changing the imageview's background color so if the image is not fit exactly, you'll see the flashing background.

[APK.zip](https://github.com/bumptech/glide/files/615740/Glide_Image_Not_Exact_Fit.zip)

[Source Code](https://github.com/xGary/sandbox)

Also, is this the proper way to downscale images? Why does Glide upscale images to the ImageView's resolution even when I am loading in a lower quality image? I am loading a bunch of images in a RecyclerView and with many phones having 2K screens, that means images are being scaled to sizes like 1440px in width.

**Comment by xGary** [comment]
Hey, sorry to bother but is there any response to this issue? Seems like every recent issue has gotten a response other than this one?

**Comment by TWiStErRob** [comment]
Hi, I'm sorry, I've been really busy lately. I do this in my free time, and haven't had much of that. Other issues got responses, because they didn't require project cloning, SDK updates, debugging, etc. Most of them can be answered by simple thinking or max code navigation. I've set up your project and was able to repro (i.e. I see the blinking colors), will get back soon.

**Comment by xGary** [comment]
Hey, no worries. Sorry, I didn't mean to rush you, just wanted to make sure this inquiry wasn't accidentally ignored.

**Comment by TWiStErRob** [comment]
Let's take an example. You ask for a `484x<ORIGINAL>` size, which displays in a `<match>x<wrap>` view, on a 1080 wide screen. Since you have `adjustViewBounds` that means the view will resize according to drawable. (Side-note: `adjustViewBounds` may also force `fitCenter`.) Glide will load a 484-wide version of the source whose size is `1920x1080`, that has an aspect of 16:9 (~1.777), so the expected Bitmap size is `484x272.25` rounded to `484x272` (your log confirms Glide loads this size). Then, this Bitmap is shown in a 1080-wide ImageView, fitted by `adjustViewBounds` to have an aspect of ~1.779 (aspect of drawable), so view size is `1080x606.94` rounded down in `ImageView.onMeasure` to `1080x606`, so the view aspect is ~1.782. So, now layout happened, and we have a drawable set as well. Here comes the drawing of the Bitmap with `fitCenter` (via `mDrawMatrix.setRectToRect` in `ImageView.configureBounds`). The matrix is set to:
```
2.2279413, 0.0,       0.8381958
0.0        2.2279413, 0.0
0.0,       0.0,       1.0
```
which means to uniformly scale by ~2.227 (fitCenter keeps aspect) and offset X by ~0.838px. Now, if you upscale the Bitmap (`484x272`) you'll get `1078.32359x606.0000336`, so you have `1.67641px` leftovers on both sides, which is reduced by the X-offset. However, since the Canvas still needs to round to whole pixels, you end up with space on the right (I guess rounding down will favor moving 1px to the left).

You can actually see this behavior better if you remove `adjustViewBounds`, because then the view height will be only 272px and you'll be able to see the left-right spacing implied by the matrix.

Since you're explicitly sizing the Glide load, and you can simply say `.fitCenter()` in code and use `scaleType="fitXY"` to fix this issue. However that's surface treatment. I think your original issue is coming from explicit sizing.

So if I read the second part of your question correctly, you want to load a 1080px wide ("low quality") image to a 1440px wide view, without having an actual 1440px wide Bitmap in memory. When Glide loads the image it cannot load a higher quality than the source image, so what you want is to keep that image, conditionally (i.e. when smaller than the target size). This behavior is `centerInside`. In #591 we had it implemented for v4. I think your easiest solution is to write your own transformation and rely on `ImageView.onDraw` to upscale it for the screen with `fitCenter`'s matrix. You can find the `CenterInside` transformation code at https://github.com/bumptech/glide/pull/727/files. If you feel like it, you can backport it to 3.x with a PR.

Hopefully taking a step back will get rid of the gap, give it a spin and let me know. Please reopen if this didn't fix the issue.

---

TL;DR
---
Gap is caused by ImageView due to rounding errors. Workaround is to use `fitXY`. Root cause treatment is to not use custom override size, but use a transformation that doesn't upscale.

**Comment by TWiStErRob** [comment]
@xGary feel free to continue conversation if you're not satisfied with the explanation above.

I closed it because I think it's working as intended due to how ImageView works.
